cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java

```diff
@@ -305,14 +315,34 @@ public void initialize(MapReduceTaskContext<Object, Object> context) throws Exce
           context, mapperMetrics, new DatasetContextLookupProvider(context), sinkPluginId);
         sink.initialize(runtimeContext);
         if (hasOneOutput) {
-          sinks.add(new SingleOutputSink<>(sink, context, runtimeContext.getMetrics()));
+          sinks.put(sinkPluginId, new SingleOutputSink<>(sink, context, runtimeContext.getMetrics()));
         } else {
-          sinks.add(new MultiOutputSink<>(sink, context, runtimeContext.getMetrics(), sinkOutputNames,
-                                          sinkOutput.getErrorDatasetName()));
+          sinks.put(sinkPluginId, new MultiOutputSink<>(sink, context, runtimeContext.getMetrics(), sinkOutputNames,
+                                                        sinkOutput.getErrorDatasetName()));
         }
+        // for sink, we will use mock metrics, as the same context is used for emitting metrics by WrappedSink
```

Comment:
This is hacky as generally Noop stuff should be used in production code. Also, passing a useless object just to fit into existing implementation makes it very confusing. I believe the class hierarchy needs to be revisit.